src/System.Text.RegularExpressions/src/System/Text/RegularExpressions/Regex.cs

```diff
@@ -1030,21 +1040,38 @@ private static CachedCodeEntry LookupCachedAndUpdate(CachedCodeEntryKey key)
         {
             lock (s_livecode)
             {
-                for (LinkedListNode<CachedCodeEntry> current = s_livecode.First; current != null; current = current.Next)
+                s_livecode.TryGetValue(key, out var entry);
+                if (entry != null )
                 {
-                    if (current.Value._key == key)
-                    {
-                        // If we find an entry in the cache, move it to the head at the same time.
-                        s_livecode.Remove(current);
-                        s_livecode.AddFirst(current);
-                        return current.Value;
-                    }
+                    CachePromoteLinkedListEntry(entry);
+                    return entry;
```

Comment:
nit, if you want you could remove this `return entry;` and change the `return null;` to `return entry;`